### PR TITLE
[547] Add coverage testing for UTxO traces

### LINKED ISSUE #547: Add coverage testing for UTXO traces
See https://github.com/input-output-hk/cardano-ledger-specs/issues/297

### PR DESCRIPTION
**Pull request by uroboros**
Closes #547 

Screenshot 2019-06-28 at 14 52 51

### CODE REVIEW

File: byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs [hunk 1]
```diff
@@ -90,3 +95,54 @@ tracesAreClassified = withTests 200 . property $ do
...
+    cover 20 "avg. nr. of tx inputs (1,5]" (1 <= avgInputs && avgInputs <= 5)
+    cover 20 "avg. nr. of tx inputs (5,10]" (5 < avgInputs && avgInputs <= 10)
+
+    cover 20 "avg. nr. of tx outputs (1,5]" (1 <= avgOutputs && avgOutputs <= 5)
+    cover 20 "avg. nr. of tx outputs (5,10]" (5 < avgOutputs && avgOutputs <= 10)
+  where
+    -- | The average "fee surplus" for transactions in the trace.
+    --   Could be zero if all the transactions had zero surplus fee.
+    avgFeeSurplus :: Trace UTXOW -> Integer -> Int
+    avgFeeSurplus tr_ n_
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs]: Why do we use the `_` suffix?

File: byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs [hunk 2]
```diff
@@ -90,3 +95,54 @@ tracesAreClassified = withTests 200 . property $ do
...
+  tr <- forAll (trace @UTXOW tl)
+  let n :: Integer
+      n = fromIntegral $ traceLength tr
+
+  when (n > 0) $ do
+    let
+        ss = preStatesAndSignals OldestFirst tr
+        txs = (body . view _2) <$> ss
+
+        nrInputs, nrOutputs :: Double
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs]: Can't we re-use these from the previous function? (I'd just take these our of the `where` clause ...

File: byron/ledger/executable-spec/test/Main.hs [hunk 3]
```diff
@@ -45,7 +45,7 @@ main = defaultMain tests
     , testGroup
       "UTxO properties"
       [ testProperty "Money is constant" moneyIsConstant
-      , testProperty "Traces are classified" UTxO.tracesAreClassified
+      , testProperty "relevant UTxO traces are generated" UTxO.relevantCasesAreCovered
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Main.hs]: We might want to capitalize the sentence, for consistency sake.

File: byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs (hunk 1 above)
Comment (uroboros) [reply byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs]: removed 👍

File: byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs (hunk 2 above)
Comment (uroboros) [reply byron/ledger/executable-spec/test/Ledger/UTxO/Properties.hs]: done, tx

File: byron/ledger/executable-spec/test/Main.hs (hunk 3 above)
Comment (uroboros) [reply byron/ledger/executable-spec/test/Main.hs]: 👍

**Pull request merged**